### PR TITLE
Add finalizer to avoid premature CRD Deletion

### PR DESCRIPTION
**Pull request by RRap0so**
## Tracking issue

Closes https://github.com/flyteorg/flyte/issues/5786

## Why are the changes needed?
To make sure Workflows aren't prematurely deleted by Flyteadmin and FlytePropeller can properly transitions the states to `Aborted`

## How was this patch tested?

Unit Tested

Tested in a staging environment manually.

## Check all the applicable boxes 

- [ ] I updated the documentation accordingly.
- [ ] All new and existing tests passed.
- [x] All commits are signed-off.

### CODE REVIEW

**Comment by pvditt** [comment]
concern if propeller never picks up the FlyteWorkflow CRD then the CRD won't be able to get garbage collected.

File: flyteadmin/pkg/workflowengine/impl/prepare_execution.go
Comment (EngHabu) [inline flyteadmin/pkg/workflowengine/impl/prepare_execution.go:132]: Doesn't this already happen here?
		SetFinalizerIfEmpty(mutableW, FinalizerKey)
https://github.com/unionai/flyte/blob/14eaa16a43bff3ada4e93c6c6421c6b35471bf30/flytepropeller/pkg/controller/handler.go#L128

Comment (pvditt) [reply flyteadmin/pkg/workflowengine/impl/prepare_execution.go:132]: @EngHabu I believe in their scenario, the CRD is getting deleted before propeller has had the chance to evaluate (and set the finalizer for) the external workflow

**Review by EngHabu — APPROVED** [review]
Pasting here from a private channel chat for anybody who comes across this:
image

I will merge after @pvditt approves.

@pvditt What's the concern with not getting deleted? if propeller is permanently down and can't act on the object then it'll be around, right?

**Comment by pvditt** [comment]
@EngHabu This simplest route is probably the best route. My concern was probably an unfounded one - the scenario would be a FlyteWorkflow CRD gets created that all propeller instances cannot access and then any TerminateExecution request wouldn't be able to delete the CRD as finalizers would be set. If this is a possibility then we can just update the TerminateExecuteRequest to include an option to ensure the CRD is deleted instead of having users utilize kubectl commands to clear a potential dangling CRD.

**Review by pvditt — APPROVED** [review]

**Pull request merged**
